math/geometry.h

```diff
@@ -358,6 +387,184 @@ bool FurthestFreePointCircle(const Eigen::Matrix<T, 2, 1>& line_start,
+                      const Eigen::Matrix<T, 2, 1>& a1,
+                      const Eigen::Matrix<T, 2, 1>& b0,
+                      const Eigen::Matrix<T, 2, 1>& b1) {
+  if (CheckLineLineCollision(a0, a1, b0, b1)) {
+    return 0;
+  }
+
+  const auto a0_proj = ProjectPointOntoLineSegment(a0, b0, b1);
+  const auto a1_proj = ProjectPointOntoLineSegment(a1, b0, b1);
+  const auto b0_proj = ProjectPointOntoLineSegment(b0, a0, a1);
+  const auto b1_proj = ProjectPointOntoLineSegment(b1, a0, a1);
+
+  const auto a0_diff_sq = (a0 - a0_proj).squaredNorm();
+  const auto a1_diff_sq = (a1 - a1_proj).squaredNorm();
+  const auto b0_diff_sq = (b0 - b0_proj).squaredNorm();
+  const auto b1_diff_sq = (b1 - b1_proj).squaredNorm();
+
+  const auto min_diff_sq =
+      std::min({a0_diff_sq, a1_diff_sq, b0_diff_sq, b1_diff_sq});
+  return std::sqrt(min_diff_sq);
+}
+
+// Check if line segment collides min and max angle on a circle, moving in a
+// counterclockwise direction from min_angle to max_angle.
+template <typename T>
+T MinDistanceLineArc(const Eigen::Matrix<T, 2, 1>& l0,
+                     const Eigen::Matrix<T, 2, 1>& l1,
+                     const Eigen::Matrix<T, 2, 1>& a_center,
+                     const T& a_radius,
+                     T a_angle_start,
+                     T a_angle_end,
+                     const int rotation_sign) {
+  a_angle_start = math_util::AngleMod(a_angle_start);
+  a_angle_end = math_util::AngleMod(a_angle_end);
+  const auto proj_center_segment =
+      ProjectPointOntoLineSegment(a_center, l0, l1);
+  const auto proj_dir_segment = proj_center_segment - a_center;
+
+  const auto l0_dir = l0 - a_center;
+  const auto l1_dir = l1 - a_center;
+
+  if (proj_dir_segment.squaredNorm() >= Sq(a_radius) ||
+      (l0_dir.squaredNorm() < Sq(a_radius) &&
+       l1_dir.squaredNorm() < Sq(a_radius))) {
+    // Out of the circle, tangent to circle, or completely in circle.
+    const T angle = math_util::AngleMod(Angle<T>(proj_dir_segment));
+    if (math_util::IsAngleBetween(
+        angle, a_angle_start, a_angle_end, rotation_sign)) {
+      // Vector of the center projected onto line segment is in arc.
+      const float distance_to_proj = (proj_dir_segment.norm() - a_radius);
+      if (distance_to_proj >= T(0)) {
+        return distance_to_proj;
+      }
+    }
+    // Vector of the center projected onto line segment is not in arc.
+    const auto a_angle_min_v = Heading(a_angle_start) * a_radius;
+    const auto a_angle_max_v = Heading(a_angle_end) * a_radius;
+    return std::sqrt(
+        std::min((proj_dir_segment - a_angle_min_v).squaredNorm(),
+                 (proj_dir_segment - a_angle_max_v).squaredNorm()));
+  }
+
+  const auto proj_center_line = ProjectPointOntoLine(a_center, l0, l1);
+  const auto proj_dir_line = proj_center_line - a_center;
+  const T along_line_dist =
+      std::sqrt(Sq(a_radius) - proj_dir_line.squaredNorm());
+
+  static const auto direction_line_intersects_arc =
```

Comment:
This should be a helper function on its own, rather than a lambda, since it has utility on its own.